### PR TITLE
Fix lxqt-leave: enter key action #1183

### PR DESCRIPTION
**Pull request by jsm222**
Fixes lxde/lxqt#1183
Solved by following https://wiki.qt.io/How_to_catch_enter_key

### CODE REVIEW

**Comment by pmattern** [comment]
Commit makes key `Enter` behave as expected in GUI `lxqt-leave`.
GTM.

(The weird behaviour of arrow keys right/left persists but is strictly speaking only an addition to lxde/lxqt#1183.)

**Comment by jsm222** [comment]
About fixing the weird behavior of arrow keys right:
I've added a [patch](https://gist.github.com/jsm222/810cd4c5fb64f66b40f6b709f96fb6a3) on top of this commit.
It makes use of the key event filter to control the arrow keys to set the QToolButtons focus appropriately in a cyclic matter. Posted there for review, and 
eventually to be committed to this pull request. (I've changed the order in buttons vector to match the gui, then I do some simple math to set next focus,passing the buttons vector to the eventfilter)

**Comment by pmattern** [comment]
Applying your patch linked [above](https://github.com/lxde/lxqt-session/pull/93#issuecomment-284433036) on top of jsm222/lxqt-session@3952462 is indeed making the arrow keys behave as expected (and leaves the fixed behaviour of key `Enter` as-is).
Pretty good!

On a side-note please let me kindly ask you to provide your contributions in dedicated branches instead of patches like the current one. This may take you a bit more time, but it helps us save a lot of time when testing, probably much more than it takes to create the branches.

File: lxqt-leave/leavedialog.cpp [hunk 1]
```diff
@@ -70,9 +70,12 @@ LeaveDialog::LeaveDialog(QWidget* parent)
         buttons.at(i)->adjustSize();
         maxWidth = qMax(maxWidth, buttons.at(i)->width());
     }
-    for (int i = 0; i < N; ++i)
-        buttons.at(i)->setMinimumWidth(maxWidth);
+    KeyEnterReceiver* key = new KeyEnterReceiver();
```
Comment (palinek) [inline lxqt-leave/leavedialog.cpp]: This object is never deleted -> leak

**Comment by palinek** [comment]
I believe processing/filtering all events here is uncessary.... I'll fiddle with the dialog a bit to try to find a simpler solution..

**Comment by palinek** [comment]
@jsm222 please, consider the #94

**Comment by jsm222** [comment]
For some reason the QToolButtons reacts to all arrow key events as if it was tab. E.g no logic in using the up/down/left/right. They all follow the tab order blindly.. I could not catch the buttons key events in your keypress event from #94. I branched my patch as suggested @pmattern 
https://github.com/jsm222/lxqt-session/tree/lxqt-leave-cyclic-arrow-keys. Please have a look. It should delete the Keyreceiver object now..

**Comment by palinek** [comment]
> They all follow the tab order blindly.. I could not catch the buttons key events in your keypress event from #94. I branched my patch as suggested @pmattern
https://github.com/jsm222/lxqt-session/tree/lxqt-leave-cyclic-arrow-keys. Please have a look. It should delete the Keyreceiver object now..

IMO the up-down navigation is not worth that hassles & tab cycling is OK

**Comment by pmattern** [comment]
Fixing the arrow keys is not so much to provide an additional means to cycle through the buttons but to avoid weird behaviour, IMO.
While I would agree that providing the tab key as only means to cycle through the buttons could be okay I find it just too weird when e. g. hitting `Down` in the middle of the first row ends in moving right.
So if there are no serious downsides I'd vote for the approach which does fix the arrow keys as well.

**Comment by palinek** [comment]
> While I would agree that providing the tab key as only means to cycle through the buttons could be okay I find it just too weird when e. g. hitting Down in the middle of the first row ends in moving right.

...and how about just making `up`, `down` keys noop?

**Comment by pmattern** [comment]
> ...and how about just making `up`, `down` keys noop?

Do you really think this would make for a good UX? Six objects are arranged in two rows and can be selected. Personally, I think arrow keys triggering moves in their respective directions is the only thing intuitive in this context.
Are all approaches which make the arrow keys behave here as everywhere else introducing some serious downsides like inadequate resource usage or why is it you seem to be so negative on these?

**Comment by agaida** [comment]
@palinek - with #94 i think this PR is obsolete now - am i right?

**Comment by palinek** [comment]
> with #94 i think this PR is obsolete now - am i right?

This also deals with arrow keys navigation, but I consider all this logic rather fragile. I've tried to redesign the leave dialog to let Qt handle the navigation -> #95

**Pull request closed without merging**